### PR TITLE
Create mocha-debug-reporter

### PR DESCRIPTION
**Pull request by scinos**
#### Changes proposed in this Pull Request

* Creates a new `mocha-debug-reporter` that will output all test info (i.e. all events detailed in https://mochajs.org/api/tutorial-custom-reporter.html) as a JSON file.
* Uses that reporter for e2e tests.

This reporter should give us more info about what (and when) hooks, tests and suites were run. As it outputs it to the filesystem it should be easy to distinguish one test from another.

#### Testing instructions

Check the e2e test artifacts contain a `mocha.log` file.

### CODE REVIEW

File: packages/mocha-debug-reporter/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	EVENT_TEST_PASS,
+} = Mocha.Runner.constants;
+
+interface Event {
+	timestamp: Date;
+	name: string;
+	data?: Record< string, unknown >;
+}
+
+export = class DebugReporter extends Mocha.reporters.Base {
```
Comment (scinos) [inline packages/mocha-debug-reporter/src/index.ts:32]: `export =` is used to generate a CJS module (https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require)

File: test/e2e/mocha-reporter.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+	"reporterEnabled": "mocha-teamcity-reporter, @automattic/mocha-debug-reporter",
+	"automatticMochaDebugReporterReporterOptions": {
```
Comment (scinos) [inline test/e2e/mocha-reporter.json:3]: `ReporterReporter` is not a typo.

Doc (https://www.npmjs.com/package/mocha-multi-reporters#configuring-reporters) says "Specify each reporter's configuration using its camel-cased name, followed by `reporterOptions`, as key. For scoped reporters such as example @myorg/custom, remove all special characters.": `automatticMochaDebugReporter` + `ReporterOptions`

**Review by WunderBart — APPROVED** [review]
Nice! 👏 🚀

File: packages/magellan-mocha-plugin/lib/test_run.js [hunk 3]
```diff
@@ -34,8 +34,7 @@ const MochaTestRun = function ( options ) {
 		basePath: path.join( options.tempAssetPath, '..', sanitizedName ),
 		counter: 0,
 	} );
-
-	process.env.TEMP_ASSET_PATH = pathWithCounter;
```
Comment (scinos) [inline packages/magellan-mocha-plugin/lib/test_run.js:38]: This code was causing a race condition: it is possible that two `MochaTestRun` gets instantiated before the corresponding `mocha` process is spawn. If this happens, both `mocha` processes will receive the _same_ value for the env var `TEMP_ASSET_PATH`.

**Comment by worldomonation** [comment]
I think this looks good, did check over the logs that were produced and it looks to be matching what the test is doing.

Now, whether half of these logs are actually useful - that's something I'd like to discuss another time 🤣

File: .teamcity/_self/projects/WebApp.kt [hunk 4]
```diff
@@ -733,8 +739,9 @@ object RunCalypsoPlaywrightE2eTests : BuildType({
 				# Install modules
 				${_self.yarn_install_cmd}
 
-				# Build package
+				# Build packages
```
Comment (worldomonation) [inline .teamcity/_self/projects/WebApp.kt:742]: Thanks for adding it here too.

**Review by worldomonation — APPROVED** [review]

**Pull request merged**